pkg/app/piped/cloudprovider/kubernetes/diff.go

```diff
@@ -75,6 +86,39 @@ func DiffList(olds, news []Manifest, opts ...diff.Option) (*DiffListResult, erro
 	return cr, nil
 }
 
+func normalizeNewSecret(old, new *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	var o, n v1.Secret
+	runtime.DefaultUnstructuredConverter.FromUnstructured(old.Object, &o)
+	runtime.DefaultUnstructuredConverter.FromUnstructured(new.Object, &n)
+
+	// Move as much as possible fields from `n.Data` to `n.StringData` to make `n` close to `o` to minimize the diff.
+	for k, v := range n.Data {
+		// Skip if the field also exists in StringData.
+		if _, ok := n.StringData[k]; ok {
+			continue
+		}
+
+		if _, ok := o.StringData[k]; !ok {
+			continue
+		}
+
+		if n.StringData == nil {
+			n.StringData = make(map[string]string)
+		}
+
+		// If the field is existing in `o.StringData`, we should move that field from `n.Data` to `n.StringData`
+		n.StringData[k] = string(v)
```

Comment:
I checked your link about Kubernetes implementation before and I didn't understand why they use that value directly but after trying using the Secret on my K8s cluster, I think we need to decode the value before assigning.
You can confirm on your cluster as well.

``` yaml
apiVersion: v1
kind: Secret
metadata:
  name: demo-secret
type: Opaque
stringData:
  username: admin
  password1: foo
data:
  password2: YmFyCg== # encoded of bar

```

What I got by `kubectl get secret demo-secret -o yaml`

``` yaml
apiVersion: v1
data:
  password: ""
  password1: Zm9v
  password2: YmFyCg==
  username: YWRtaW4=
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"password2":"YmFyCg=="},"kind":"Secret","metadata":{"annotations":{},"name":"demo-secret","namespace":"default"},"stringData":{"password1":"foo","username":"admin"},"type":"Opaque"}
  creationTimestamp: "2022-02-28T05:57:59Z"
  name: demo-secret
  namespace: default
  resourceVersion: "326657111"
  uid: e101fa74-152e-4f6d-9c15-3f3d9ab95d61
type: Opaque
```

Reply:
Thank you. I'll try it too.

Reply:
The first one is simply apply, and second one is using piped.
The data looks like same, so I think we do not need to decode before assigning.
```
 k get secret demo-secret -o yaml
apiVersion: v1
data:
  password1: Zm9v
  password2: YmFyCg==
  username: YWRtaW4=
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"password2":"YmFyCg=="},"kind":"Secret","metadata":{"annotations":{},"name":"demo-secret","namespace":"pipecd"},"stringData":{"password1":"foo","username":"admin"},"type":"Opaque"}
  creationTimestamp: "2022-02-28T06:23:11Z"
  name: demo-secret
  namespace: pipecd
  resourceVersion: "28265696"
  uid: ace4e264-a5dd-4824-8202-694490536b2a
type: Opaque
```
```
apiVersion: v1
data:
  password1: Zm9v
  password2: YmFyCg==
  username: YWRtaW4=
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"password2":"YmFyCg=="},"kind":"Secret","metadata":{"annotations":{"pipecd.dev/application":"eeae5233-0c01-40b0-8c42-910edefd2aa0","pipecd.dev/commit-hash":"56615e47bb8723223154ca102c782b8525e5d6bf","pipecd.dev/managed-by":"piped","pipecd.dev/original-api-version":"v1","pipecd.dev/piped":"d90c7d7f-80ca-4684-93ad-bbab1f2e5f99","pipecd.dev/resource-key":"v1:Secret:default:demo-secret","pipecd.dev/variant":"primary"},"name":"demo-secret","namespace":"default"},"stringData":{"password1":"foo","username":"admin"},"type":"Opaque"}
    pipecd.dev/application: eeae5233-0c01-40b0-8c42-910edefd2aa0
    pipecd.dev/commit-hash: 56615e47bb8723223154ca102c782b8525e5d6bf
    pipecd.dev/managed-by: piped
    pipecd.dev/original-api-version: v1
    pipecd.dev/piped: d90c7d7f-80ca-4684-93ad-bbab1f2e5f99
    pipecd.dev/resource-key: v1:Secret:default:demo-secret
    pipecd.dev/variant: primary
  creationTimestamp: "2022-02-28T06:46:17Z"
  name: demo-secret
  namespace: default
  resourceVersion: "28277530"
  uid: 7c3e3729-37bf-4aa7-9006-c7ffec0a2041
type: Opaque
```

Reply:
@Hosshii I think we don't need to compare between the applied data from `piped` and `kubectl`.
What we need to compare is the Secret before and after applying.
In `normalizeNewSecret` function, `old` is the raw manifest data from Git (the one before applying), and `new` the data stored in K8s' database (the one can get by `kubectl get`).

For example, I have this manifest in Git `secret.yaml`

``` yaml
apiVersion: v1
kind: Secret
metadata:
  name: demo-secret
type: Opaque
stringData:
  username: admin
  password1: foo
data:
  password2: YmFyCg== # encoded of bar
```
(This will be the `old` parameter of the `normalizeNewSecret` function)

And I apply it with `kubectl apply -f secret.yaml`.
Then I retrieve it stored data with `kubectl get secret demo-secret -o yaml` and I will have

``` yaml
apiVersion: v1
data:
  password1: Zm9v
  password2: YmFyCg==
  username: YWRtaW4=
kind: Secret
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","data":{"password2":"YmFyCg=="},"kind":"Secret","metadata":{"annotations":{},"name":"demo-secret","namespace":"default"},"stringData":{"password1":"foo","username":"admin"},"type":"Opaque"}
  creationTimestamp: "2022-02-28T05:57:59Z"
  name: demo-secret
  namespace: default
  resourceVersion: "326657111"
  uid: e101fa74-152e-4f6d-9c15-3f3d9ab95d61
type: Opaque
```

(This will be the `new` parameter of the `normalizeNewSecret` function)

Let's take a look at the `password1` field, its data has been base64 encoded to `Zm9v`.

That is why I said that we have to decode before assigning to compare.

Reply:
Thank you. This is handled by `runtime.DefaultUnstructuredConverter.FromUnstructured`(converted from base64 to string in [here](https://github.com/kubernetes/kubernetes/blob/9968b0eff06f109792e64f37ee92889711713a88/staging/src/k8s.io/apimachinery/pkg/runtime/converter.go#L462)) and `runtime.DefaultUnstructuredConverter.ToUnstructured`. So I think we don't have to decode ourselves.

Reply:
Got it. So let;s merge this.